scipy/sparse/_dok.py

```diff
@@ -274,19 +357,40 @@ def _mul_scalar(self, other):
         return new
 
     def _matmul_vector(self, other):
-        # matrix * vector
-        result = np.zeros(self.shape[0], dtype=upcast(self.dtype, other.dtype))
+        res_dtype = upcast(self.dtype, other.dtype)
+
+        # vector @ vector
+        if self.ndim == 1:
+            if issparse(other):
+                if other.format == "dok":
+                    shared_keys = self.keys() & other.keys()
+                else:
+                    shared_keys = self.keys() & other.tocoo().coords[0]
+                arr = np.array([self._dict[k] * other._dict[k] for k in shared_keys])
```

Comment:
Can we do: `sum(self._dict[k] * other._dict[k] for k in shared_keys)` and save some temporary allocations?